proposals/XXXX-build-time-constant-values.md

```diff
@@ -0,0 +1,188 @@
+
+## Proposed Solution
+
+The Swift compiler will recognize declarations of properties, local variables, and function parameters declared with a `@const` attribute as having an additional guarantee to be known at compile-time. If a `@const` property or variable is initialized with a runtime value, the compiler will emit an error. Similarly, if a runtime value is passed as an argument to a `@const` function parameter, the compiler will emit an error. Aside from participating in name mangling, the attribute has no runtime effect.
+
+## Detailed Design
+
+### Property `@const` attribute
+
+A stored property on a `struct` or a `class` can be marked with a `@const` attribute to indicate that its value is known at compile-time. 
+
+```
+struct Foo {
+  @const let title: String = "foo"
+}
+```
+
+Similarly to Implicitly-Unwrapped Optionals, the mental model for semantics of this attribute is that it is a flag on the declaration that  guarantees that the compiler is able to know its value as shared by all instance of the type. For now, `@const let` and `@const static let` are equivalent in what information the `@const` attribute conveys to the compiler.(*AC: Worth mentioning here that in the future,* `@const let` *properties will gain an ability to be initialized with compile-time expressions*). Unlike a plain `static let` stored property, a `@const static let` property does not need to refer to a location in memory shared by all instances of the type, and can be elided by the compiler entirely. Default-initializing a `@const` property with a runtime value or not default-initializing it at all results in an compilation error.
+
+### Parameter `@const` attribute
+
+A function parameter can be marked with a `@const`  keyword to indicate that values passed to this parameter at the call-site must be compile-time-known values.
+
+```
+func foo(input: @**const** Int) {...}
+```
+
+### Supported Types
+
+The requirement that values of `@const` properties and parameters be known at compile-time restricts the allowable types for such declarations. The current scope of the proposal includes:
+
+* Integer and floating-point literals.
+* String literals (excluding interpolated strings).
+* Enum cases with no associated values.
+* Array literals consisting of integer, floating-point, string literals or enum cases with no associated values.
+* Dictionary literals consisting of integer, floating-point, string literals or enum cases with no associated values.
+
+This list may expand in the future to include more literal-value kinds or potential new compile-time valued constructs.
```

Comment:
Can probably say "will" expand. Could even give an example of @frozen structs with public members.